src/scripts/iapSrv.js

```diff
@@ -1,270 +1,648 @@
-                                currentStoreProducts.push({
-                                    id: productId,
-                                    type: productType,
-                                    alias: productAlias
-                                });
-                                IapSrv.subscriptionLengthInMonth[productId] = product.length;
-                            }
+                    console.log('starting purchase');
+
+                    return iapStoreReadyProm.then(function(){
+                        console.log('purchase: isPurchaseInProgress=' + iapSrv.isPurchaseInProgress);
+                        if (iapSrv.isPurchaseInProgress){
+                            console.log('purchase is already in progress');
+                            return $q.reject(false);
+                        }
+                        iapSrv.isPurchaseInProgress = true;
+                        iapSrv.purchaseInProgressDfd = $q.defer();
+                        $ionicLoading.show({
+                            template: 'Purchase is in progress...'
                         });
 
-                        IapSrv.productIds = productsIdsArr;
-                        return currentStoreProducts;
-                    }).then(function (currentStoreProductsArr) {
-                        _store = $window.store;
+                        if (enableNoStoreMode || isWeb){
+                            var validator = _getValidatorFunc();
+                            var mockProductForWeb = {};
+                            var appProduct = iapSrv.getAppProduct(productId);
 
-                        // Enable maximum logging level
-                        _store.verbosity = ENV.debug ? _store.DEBUG : _store.QUIET;
+                            angular.extend(mockProductForWeb, appProduct, extendedProductMock);
 
-                        // Enable remote receipt validation
-                        _store.validator = function (product, callback) {
-                            var validateProm = InAppPurchaseHelperSrv.validateReceipt(product.transaction.appStoreReceipt);
-                            validateProm.then(function (res) {
-                                callback(res);
+                            validator(mockProductForWeb).then(function(res){
+                                if (res){
+                                    console.log('mock purchase completed');
+                                    iapSrv.purchaseInProgressDfd.resolve(appProduct);
+                                    iapSrv.isPurchaseInProgress = false;
+                                    $ionicLoading.hide();
+                                    console.log('purchase: isPurchaseInProgress=' + iapSrv.isPurchaseInProgress);
+                                }
+                                else{
+                                    console.log('error in validating mock purchase');
+                                    iapSrv.purchaseInProgressDfd.reject();
```

Comment:
Finish purchase function can be used instead of repeating those 3 line(iapSrv.purchaseInProgressDfd,iapSrv.isPurchaseInProgress,iapSrv.isPurchaseInProgress)